src/libraries/System.Linq/src/System/Linq/SegmentedArrayBuilder.cs

```diff
@@ -0,0 +1,313 @@
+        /// <summary>The total number of elements in all but the current/last segment.</summary>
+        private int _countInFinishedSegments;
+        /// <summary>The number of elements in the current/last segment.</summary>
+        private int _countInCurrentSegment;
+
+        /// <summary>Initialize the builder.</summary>
+        /// <param name="scratchBuffer">A buffer that can be used as part of the builder.</param>
+        public SegmentedArrayBuilder(Span<T> scratchBuffer)
+        {
+            _currentSegment = _firstSegment = scratchBuffer;
+        }
+
+        /// <summary>Clean up the resources used by the builder.</summary>
+        public void Dispose()
+        {
+            int segmentsCount = _segmentsCount;
+            if (segmentsCount != 0)
+            {
+                ReadOnlySpan<T[]> segments = _segments;
+
+                // We need to return all rented arrays to the pool, and if the arrays contain any references,
+                // we want to clear them first so that the pool doesn't artificially root contained objects.
+                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+                {
+                    // Return all but the last segment. All of these are full and need to be entirely cleared.
+                    foreach (T[] segment in segments.Slice(0, segmentsCount - 1))
+                    {
+                        ArrayPool<T>.Shared.Return(segment, clearArray: true);
+                    }
+
+                    // For the last segment, we can clear only what we know was used.
+                    T[] currentSegment = segments[segmentsCount - 1];
+                    Array.Clear(currentSegment, 0, _countInCurrentSegment);
+                    ArrayPool<T>.Shared.Return(currentSegment);
+                }
+                else
+                {
+                    // Return every rented array without clearing.
+                    foreach (T[] segment in segments.Slice(0, segmentsCount))
+                    {
+                        ArrayPool<T>.Shared.Return(segment);
+                    }
+                }
+            }
+        }
+
+        /// <summary>Gets the number of elements in the builder.</summary>
+        public readonly int Count => checked(_countInFinishedSegments + _countInCurrentSegment);
+
+        /// <summary>Adds an item into the builder.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Add(T item)
+        {
+            Span<T> currentSegment = _currentSegment;
+            int countInCurrentSegment = _countInCurrentSegment;
+            if ((uint)countInCurrentSegment < (uint)currentSegment.Length)
+            {
+                currentSegment[countInCurrentSegment] = item;
+                _countInCurrentSegment++;
+            }
+            else
+            {
+                AddSlow(item);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
Nit: extract to local method to emphasise that it isn't being used elsewhere?

Reply:
Ok

Reply:
Ah, no, I can't. This is a struct.